data-generation/src/main/scala/thesis/SnapshotDelta.scala

```diff
@@ -53,23 +69,97 @@ class SnapshotDelta(val graphs: MutableList[Snapshot],
 
   override def snapshotAtTime(instant: Instant): AttributeGraph = {
 
-    val closestGraph = getClosestGraph(graphs, instant)
-    logger.warn(s"Instant $instant, Closest :graph${closestGraph.instant}")
-    if (closestGraph.instant == instant) {
-      logger.warn("The queried graph is already materialized")
-      closestGraph.graph
-    } else if (instant < closestGraph.instant) {
-      logger.warn("Closest graph in the future, and there is no one in the past.")
-      val initialLogs = getLogsInInterval(logs, Interval(Instant.MIN, instant))
-      createGraph(initialLogs)
-    } else {
-      logger.warn("Closest graph is in the past. Need to apply logs forwards")
-      val logsToApply = getLogsInInterval(logs, Interval(closestGraph.instant, instant))
-      forwardApplyLogs(closestGraph.graph, logsToApply)
+    val mostRecentMaterializedSnapshot = getMostRecentMaterializedSnapshot(graphs, instant)
+    mostRecentMaterializedSnapshot match {
+      case Some(snapshot) =>
+        if (snapshot.instant == instant) {
+          logger.warn("The queried graph is already materialized")
+          snapshot.graph
+        } else {
+          logger.warn("Closest graph is in the past. Need to apply logs forwards")
+          val logsToApply = getLogsInInterval(logs, Interval(snapshot.instant, instant))
+          forwardApplyLogs(snapshot.graph, logsToApply)
+        }
+      case None =>
+        logger.warn("Closest graph in the future, and there is no one in the past.")
+        val initialLogs = getLogsInInterval(logs, Interval(Instant.MIN, instant))
+        createGraph(initialLogs)
     }
   }
-  
+
   override def directNeighbours(vertexId: VertexId, interval: Interval): RDD[VertexId] = throw new NotImplementedError()
+
+  /** get Vertex at a certain point in time
+   *
+   * This method shares a lot of functionality with getting a snapshot at a specific time,
+   * but only for a single vertex. A lot has been duplicated because we try to only include relevant
+   * logs and save processing with this. (Through benchmarking its roughly 60% faster than just using
+   * snapshotAtTime() and filtering the specific vertex)
+   *
+   * @param vertex  vertex to be found
+   * @param instant that specific time
+   * @return Possibly vertex if it exists at the time
+   */
+  def getVertex(vertex: VERTEX, instant: Instant): Option[(Entity, Attributes)] = {
+    val possibleMaterializedSnapshot = getMostRecentMaterializedSnapshot(graphs, instant)
+    possibleMaterializedSnapshot match {
+      case Some(snapshot) => getMaterializedVertex(vertex, instant, snapshot)
+      case None => getUnmaterializedVertex(vertex, instant)
+    }
+  }
+
+  // Thinking about moving these two functions to another file. Thoughts?
+  def getMaterializedVertex(vertex: VERTEX, instant: Instant, snapshot: Snapshot): Option[(VERTEX, Attributes)] = {
+    val Snapshot(graph, mInstant) = snapshot
```

Comment:
```suggestion
    val Snapshot(graph, materializedInstant) = snapshot
```
Siden det er tanken ja? Synes personlig dette er mer lesbart, men up to you